File: app/src/main/java/com/garymcgowan/moviepedia/network/OmdbMoviesAPI.java
package com.garymcgowan.moviepedia.network;

import com.garymcgowan.moviepedia.model.Movie;

import io.reactivex.Flowable;
import io.reactivex.Single;
import retrofit2.http.GET;
import retrofit2.http.Query;

/**
 * Created by Gary on 2016/11/09.
 */

/**
 * api: http://www.omdbapi.com/
 */
public interface OmdbMoviesAPI {

    String API_KEY = "955c021d"; // TODO this can be extracted to gradle property

    String TYPE_MOVIE = "movie";
    String TYPE_SERIES = "series";
    String TYPE_EPISODE = "episode";

    String PLOT_SHORT = "short";
    String PLOT_FULL = "full";

    String RESPONSE_JSON = "json";
    String RESPONSE_XML = "xml";

    String TRUE = "true";
    String FALSE = "false";

    @GET("/?apikey=" + API_KEY)
    Flowable<OmdbSearch> getObservableMoviesSearch(
            @Query("s") String search,
            @Query("type") String type,// movie, series, episode
            @Query("y") Integer year,
            @Query("r") Boolean responseType,//json, xml
            @Query("page") Integer page,
            @Query("callback") String callback,
            @Query("v") Integer apiVersion);

    @GET("/?apikey=" + API_KEY)
    Single<Movie> getObservableMovie(
            @Query("i") String id,
            @Query("t") String title,
            @Query("type") String type,// movie, series, episode
            @Query("y") Integer year,
            @Query("plot") String plotLength,//short, full
            @Query("r") Boolean responseType,//json, xml
            @Query("tomatoes") Boolean tomatoes,
            @Query("callback") String callback,
            @Query("v") Integer apiVersion);

}

File: app/src/main/java/com/garymcgowan/moviepedia/network/OmdbMoviesAPI.kt
package com.garymcgowan.moviepedia.network

import com.garymcgowan.moviepedia.model.Movie

import io.reactivex.Flowable
import io.reactivex.Single
import retrofit2.http.GET
import retrofit2.http.Query

/**
 * api: http://www.omdbapi.com/
 */
interface OmdbMoviesAPI {

    @GET("/?apikey=$API_KEY")
    fun getObservableMoviesSearch(
            @Query("s") search: String,
            @Query("type") type: String?, // movie, series, episode
            @Query("y") year: Int?,
            @Query("r") responseType: String, //json, xml
            @Query("page") page: Int?,
            @Query("callback") callback: String?,
            @Query("v") apiVersion: Int?): Flowable<OmdbSearch>

    @GET("/?apikey=$API_KEY")
    fun getObservableMovie(
            @Query("i") id: String,
            @Query("t") title: String?,
            @Query("type") type: String?, // movie, series, episode
            @Query("y") year: Int?,
            @Query("plot") plotLength: String, //short, full
            @Query("r") responseType: String, //json, xml
            @Query("tomatoes") tomatoes: Boolean?,
            @Query("callback") callback: String?,
            @Query("v") apiVersion: Int?): Single<Movie>

    companion object {

        const val API_KEY = "955c021d" // TODO this can be extracted to gradle property

        const val TYPE_MOVIE = "movie"
        const val TYPE_SERIES = "series"
        const val TYPE_EPISODE = "episode"

        const val PLOT_SHORT = "short"
        const val PLOT_FULL = "full"

        const val RESPONSE_JSON = "json"
        const val RESPONSE_XML = "xml"

        const val TRUE = "true"
        const val FALSE = "false"
    }

}
